### PR TITLE
Avoid confounding compilation side effects of `@time_imports`

### PR DESCRIPTION
**Pull request by IanButterworth**
Makes functions for the report printing that can be precompiled into the sysimage.

### CODE REVIEW

File: contrib/generate_precompile.jl [hunk 1]
```diff
@@ -181,6 +205,7 @@ julia_exepath() = joinpath(Sys.BINDIR, Base.julia_exename())
 Artifacts = get(Base.loaded_modules,
           Base.PkgId(Base.UUID("56f22d72-fd6d-98f1-02f0-08ddc0907c33"), "Artifacts"),
           nothing)
+# use this as an opportunity to
```
Comment (giordano) [inline contrib/generate_precompile.jl:208]: Incomplete comment?

File: contrib/generate_precompile.jl [hunk 2]
```diff
@@ -114,6 +114,30 @@ precompile(Tuple{typeof(Base.setup_stdio), Base.TTY, Bool})
...
 precompile(Tuple{Type{Base.UUID}, Base.UUID})
+
+# @time and @time_imports
+precompile(Tuple{typeof(Base.Threads.lock_profiling), Bool})
+precompile(Tuple{typeof(Base.getindex), Base.Threads.Atomic{Int}})
+precompile(Tuple{typeof(Base.gc_num)})
+precompile(Tuple{typeof(Base.cumulative_compile_timing), Bool})
+precompile(Tuple{typeof(Base.cumulative_compile_time_ns)})
+precompile(Tuple{typeof(Base.Threads.atomic_add!), Base.Threads.Atomic{Int}, Int})
+precompile(Tuple{Base.var"##printstyled#876", Bool, Bool, Bool, Bool, Bool, Bool, Symbol, typeof(Base.printstyled), String, Vararg{String}})
```
Comment (KristofferC) [inline contrib/generate_precompile.jl:125]: This is likely to get stale and break the build.

File: contrib/generate_precompile.jl [hunk 3]
```diff
@@ -114,6 +114,30 @@ precompile(Tuple{typeof(Base.setup_stdio), Base.TTY, Bool})
...
+precompile(Tuple{typeof(Base.Threads.atomic_sub!), Base.Threads.Atomic{Int}, Int})
+precompile(Tuple{Type{Base.GC_Diff}, Base.GC_Num, Base.GC_Num})
+precompile(Tuple{typeof(Base.getproperty), Base.GC_Diff, Symbol})
+precompile(Tuple{Type{NamedTuple{(:value, :time, :bytes, :gctime, :gcstats, :lock_conflicts, :compile_time, :recompile_time), T} where T<:Tuple}, Tuple{Nothing, Float64, Int, Float64, Base.GC_Diff, Int, Float64, Float64}})
+precompile(Tuple{typeof(Base.getproperty), NamedTuple{(:value, :time, :bytes, :gctime, :gcstats, :lock_conflicts, :compile_time, :recompile_time), Tuple{Nothing, Float64, Int, Float64, Base.GC_Diff, Int, Float64, Float64}}, Symbol})
+precompile(Tuple{typeof(Base.:(*)), Float64, Float64})
+precompile(Tuple{typeof(Base.gc_alloc_count), Base.GC_Diff})
+precompile(Tuple{typeof(Core.kwcall), NamedTuple{(:msg,), Tuple{Nothing}}, typeof(Base.time_print), Base.TTY, Float64, Int, Int, Int, Int, Float64, Float64, Bool})
+precompile(Tuple{Base.var"#865#866"{Nothing, Float64, Int, Int, Int, Float64, Float64, Bool, String}, Base.GenericIOBuffer{Memory{UInt8}}})
```
Comment (KristofferC) [inline contrib/generate_precompile.jl:138]: Same here

**Comment by IanButterworth** [comment]
@KristofferC I figured breaking the printing out into functions that can be precompiled into the sysimage makes more sense

**Comment by IanButterworth** [comment]
Also tidies up the interleaved printing with `--trace-compile` so that it's blocks of precompile between each report line making it easier to understand what caused them.
```
% ./julia --start=no --trace-compile=stderr --trace-compile-timing -e "using InteractiveUtils; @time @time_imports using MicroMamba"
      0.5 ms  Printf
     15.0 ms  Dates
      0.3 ms  Scratch
     10.0 ms  LazyArtifacts
      1.0 ms  TOML
      7.2 ms  Preferences
      0.5 ms  JLLWrappers
#=    0.0 ms =# precompile(Tuple{typeof(micromamba_jll.__init__)})
#=    0.0 ms =# precompile(Tuple{typeof(micromamba_jll.find_artifact_dir)})
#=    2.4 ms =# precompile(Tuple{Type{Base.Val{x} where x}, Module})
#=    2.0 ms =# precompile(Tuple{Type{NamedTuple{(:honor_overrides,), T} where T<:Tuple}, Tuple{Bool}})
#=  439.9 ms =# precompile(Tuple{typeof(Artifacts._artifact_str), Module, String, Base.SubString{String}, String, Base.Dict{String, Any}, Base.SHA1, Base.BinaryPlatforms.Platform, Base.Val{LazyArtifacts}})
#=   24.4 ms =# precompile(Tuple{typeof(Base.unique!), Array{String, 1}})
#=    1.3 ms =# precompile(Tuple{typeof(Base.invokelatest), Any})
#=    0.0 ms =# precompile(Tuple{typeof(JLLWrappers.get_julia_libpaths)})
#=    7.8 ms =# precompile(Tuple{typeof(Base.vcat), Array{String, 1}, Array{String, 1}})
               ┌ 545.2 ms micromamba_jll.__init__() 99.93% compilation time (88% recompilation)
    545.7 ms  micromamba_jll 99.84% compilation time (88% recompilation)
      0.4 ms  MicroMamba
#=    2.0 ms =# precompile(Tuple{Type{NamedTuple{(:value, :time, :bytes, :gctime, :gcstats, :lock_conflicts, :compile_time, :recompile_time), T} where T<:Tuple}, Tuple{Nothing, Float64, Int64, Float64, Base.GC_Diff, Int64, Float64, Float64}})
#=    2.9 ms =# precompile(Tuple{typeof(Base.getproperty), NamedTuple{(:value, :time, :bytes, :gctime, :gcstats, :lock_conflicts, :compile_time, :recompile_time), Tuple{Nothing, Float64, Int64, Float64, Base.GC_Diff, Int64, Float64, Float64}}, Symbol})
#=   26.4 ms =# precompile(Tuple{typeof(Core.kwcall), NamedTuple{(:msg,), Tuple{Nothing}}, typeof(Base.time_print), Base.TTY, Float64, Int64, Int64, Int64, Int64, Float64, Float64, Bool})
  0.586889 seconds (763.19 k allocations: 40.931 MiB, 5.56% gc time, 92.84% compilation time: 88% of which was recompilation)
#=   13.0 ms =# precompile(Tuple{Base.var"#start_profile_listener##2#start_profile_listener##3"{Task, Base.AsyncCondition}})
```

**Comment by DilumAluthge** [comment]
This doesn't backport cleanly to 1.10.

@IanButterworth Would you be able to make a manual backport to 1.10?

**Pull request merged**